### DISCUSSION TITLE
Add virtual display driver

### DISCUSSION THREAD

**Discussion by gzxiexl** [LizardByte/community#230] [Sunshine]
Hope to support virtual display if no other displays are present.

**Comment by emiksel** [comment]
on point. anyone found any solution?

**Reply by Scarfmonster** [reply]
I'm currently using [IddSampleDriver](https://github.com/ge9/IddSampleDriver) on Windows. so far I haven't had any issues with it.

**Reply by F-TD5X** [reply]
and also [usbmmidd](https://www.amyuni.com/downloads/usbmmidd_v2.zip) , actually did same thing but store settings in registry

**Reply by the-V01D** [reply]
no, not just if no displays are present. add an option that adds a virtual display when a device connects in settings. I'm using Sunshine as a second monitor because of its low latency.

**Reply by DZXTRICKS** [reply]
How to use the IddSampleDriver with Sunshine? i tried changing the Output Name to Display2 (based on windows display setting showed me), but then it output  ddprobe.exe [1] [\\.\DISPLAY2] returned: 0x00000000 There must be something i configured wrong

Edit: so apparently i had to use dxgi tool to see what display it actually is. The tool tell me instead of Display2 like Windows Display Setting shows, it was Display4. No wonder it failed to show anything

**Comment by gzxiexl** [comment]
https://github.com/pavlobu/deskreen/discussions/86#discussioncomment-7119465

share two project
https://github.com/nomi-san/parsec-vdd
Standalone ParsecVDD, create a virtual super display without Parsec, upto 4K 2160p@240hz.
https://github.com/KtzeAbyss/Easy-Virtual-Display
Create virtual displays in Windows with ease, supporting a range of resolutions and refresh rates (such as 4K 240Hz). Ideal for remote control or graphics card spoofing.

**Reply by jimmt41** [reply]
Gzxiexl, these are great projects! I was looking for something that would automatically switch primary to the VD for remote streaming.

**Reply by buthig666** [reply]
Looking forward to Moonlight advertising Res/refresh-rate/HDR condition and Sunshine responding with a virtual desktop condition for it.  The dream.  Something similar to what this guy is charging for from forking Sunshine/moonlight. https://github.com/DuoStream/Duo

**Comment by dingyifei** [comment]
> Gzxiexl, these are great projects! I was looking for something that would automatically switch primary to the VD for remote streaming.
@gzxiexl 

I'm also trying to get a virtual display so my trimonitor setup wouldn't waste power when I'm not home.

I switch my monitors off by using displayswitch.exe command with the option "internal" so my additional displays are in power saving mode.

 Displayswitch.exe probably works for your problem (Switch monitors to clone mode). 

https://ss64.com/nt/displayswitch.html

**Comment by Dampfwalze** [comment]
Just want to throw this in here: 

https://github.com/MolotovCherry/virtual-display-rs is a highly configurable virtual display driver written in Rust. It supports multiple virtual monitors and is configurable on the fly using IPC, Windows Pipes in this case. The Plug and Play version comes with a nice GUI Tool to configure it, while it is running. I think this driver can be integrated directly into Sunshine really nicely. This is a very fresh Project, with its first commit just 6 month ago, but seems to be very complete already!

**Reply by devxan** [reply]
How can this be integrated directly into Sunshine really well?

**Reply by Dampfwalze** [reply]
Sunshine can just talk to it over a named pipe and send commands to change it's behavior. In fact there is an ongoing discussion about building a library around it's IPC protocol that can be linked into any application. For Sunshine, it is just a metter of distributing the driver and designing the logic on how and when to create a virtual monitor around it.

**Comment by gongfuture** [comment]
My solution is using ```usbmmidd_v2```

Then add three global commands
1. ```cmd /c reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\WUDF\Services\usbmmIdd\Parameters\Monitors" /v 0 /t REG_SZ /d %SUNSHINE_CLIENT_WIDTH%,%SUNSHINE_CLIENT_HEIGHT% /f```
2. ```cmd /c reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\WUDF\Services\usbmmIdd\Parameters\Monitors" /ve /t REG_SZ /d %SUNSHINE_CLIENT_WIDTH%,%SUNSHINE_CLIENT_HEIGHT% /f```
3. Do ```"C:\path\usbmmidd_v2\deviceinstaller64.exe" enableidd 1```   Undo ```"C:\path\usbmmidd_v2\deviceinstaller64.exe" enableidd 0```

All the three commands run as admin, and set the virtual display as the main display

**Reply by naddel91** [reply]
I just added those three commands to "Apps" under sunshine, but I cannot start it from moonlight. What exactly am I missing?

**Reply by gongfuture** [reply]
These three commands should be put under "Config"-"General"-"Command Preparations".
![image](https://github.com/LizardByte/Sunshine/assets/60888755/71ed0c7d-c681-4b46-92f8-116399a747c1)
And if you don't want some of the apps create a virtual display, such as using "Desktop", just remember to turn "Global Prep Commands" off in your application config.
![image](https://github.com/LizardByte/Sunshine/assets/60888755/561fffdd-50c8-4137-811c-6216dd4f2ffe)

**Reply by DZXTRICKS** [reply]
Have you checked the app dxgi tool that is inside Sunshine folder? You can use a cmd to run it. Check if your virtual driver has running or not from there. Note the specified number or ID. Then, put that onto the sunshine config. (That's what worked from me, I just didn't check first)

**Reply by naddel91** [reply]
![dxgi](https://github.com/LizardByte/Sunshine/assets/83251379/de74f1f7-7016-4181-b11a-f291f457dac5)

That is what I have after installing virtual device "usbmmidd_v2"

**Reply by naddel91** [reply]
copied exactly. then I get error -1 when trying to boot this app in moonlight from my android phone.

**Reply by DZXTRICKS** [reply]
Just like yours, my display is noted to be DISPLAY144, yours happened to be 162. Just use it as a display input for sunshine in the config. After doing that it just works on my Android tablet.

If there is still an issue, like black screen or encode error. Last time the issue for me was using my dedicated GPU as encoder but the virtual display is apparently running using my integrated GPU. So I had to switch encoder to integrated GPU.

**Reply by naddel91** [reply]
![display](https://github.com/LizardByte/Sunshine/assets/83251379/eaec22aa-ea6a-4c22-b7c6-9aabb075b3f2)

there you go. but then I open the usbmmidd_v2 app in moonlight, my main monitor flickers a few times and then I get an error in moonlight. I do not have an integrated GPU on my ryzen 5 5600. that should not be the problem here.

**Reply by DZXTRICKS** [reply]
Sadly as far as I know "Microsoft basic render device" can't be set into encoder in sunshine. So it has to be your XT somehow. Maybe your virtual display needs to be configured so it runs using the XT's driver? I don't know at this point  (and yes you have a point with that iGPU missing in Ryzen 5 5600, good choice of hardware if I should say myself 👍🏻, good values)

**Reply by naddel91** [reply]
I can however start a random game, drag it onto the virtual display manually and then turn off BOTH main monitors in order to make it work. but that is not a solution when I am not at home. goal is to achieve remote play using only moonlight and not having to manually drag game windows onto a virtual display.

**Reply by naddel91** [reply]
I can start games now with monitors OFF, but moonlight shows the main monitor instead of the virtual one where the game should be running. 

so I am "only" missing one thing.
either:
1. a way to switch monitors in moonlight on the fly? 
or 2. a solution to press "Win"+"Shift"+"arrow key" to switch the running application to another monitor.

is that possible here in sunlight/moonlight?

**Reply by Darkhost1999** [reply]
Enable the touch keyboard on your taskbar inside Windows and you can "Win"+"Shift"+"arrow key"
I always boot desktop with moonlight never directly into a game.

**Comment by ClassicOldSong** [comment]
Just wrote a little helper to start new virtual display automatically when moonlight client connects, matching client's resolution and framerate settings - just like what Sidecar does

https://github.com/ClassicOldSong/Widecar

**Reply by madmalkav** [reply]
This is so cool,  I'm setting this right now, thanks for your work.

PS: anyone knows if something like this exists for Linux/Wayland?
